Awesome Endeavour: Circuit Relay

dryajov opened the pull request:
We're getting close to completing the JS circuit implementation, this issue is to track further progress. Main work is being done here - ~~https://github.com/libp2p/js-libp2p-circuit/pull/9~~ https://github.com/libp2p/js-libp2p-circuit/pull/14.

### Assumptions (needs validation)
   1. All addresses are encapsulated with `p2p-circuit` by default to make them dialable over circuit relay
   2. If an explicit relay address is present don't add defaults, use that as the address instead
   3. Out of all available transports relay is dialed last

### Outstanding work (this will be kept up to date with specific tasks as they come up):
* [x] Complete the relay spec https://github.com/libp2p/specs/pull/15
* [x] Verify and complete integration points in `js-ipfs` and `libp2p`
    - [x] Finalize integrating with `libp2p` 
    - [x] Finalize integrating with `js-ipfs`
* [ ] More test coverage and integration tests (WIP)
    - this is the latest coverage report for unit tests, lots more is needed but we're getting there

```
=============================== Coverage summary ===============================
Statements   : 71.03% ( 255/359 )
Branches     : 54.96% ( 72/131 )
Functions    : 70.83% ( 17/24 )
Lines        : 71.23% ( 250/351 )
================================================================================
```
* [x] Cleanup implementation
   - [x] cleanup pull stream read/write logic
   - [x] cleanup `relay.js`
   - [x] cleanup `listener.js`

### Outstanding PRs:

| Name | Ready for Review | Reviewed | Merged | Published | 
|-------|-------------------|---------|---------|-----------|
|[libp2p-tcp](https://github.com/libp2p/js-libp2p-tcp/pull/81) |✅|✅|✅|✅|
|[libp2p-websockets](https://github.com/libp2p/js-libp2p-websockets/pull/62) |✅|✅|✅|✅|
|[libp2p-webrtc-star](https://github.com/libp2p/js-libp2p-webrtc-star/pull/121) |✅|✅|✅|✅|
|[libp2p-webrtc-direct](https://github.com/libp2p/js-libp2p-webrtc-direct/pull/15) |✅|✅|✅|✅|
|[libp2p-circuit](https://github.com/libp2p/js-libp2p-circuit/pull/14) |✅|✅|✅|✅|
|[libp2p-swarm](https://github.com/libp2p/js-libp2p-swarm/pull/224/files)|✅|✅|✅|✅|
|[libp2p](https://github.com/libp2p/js-libp2p/pull/116) |✅|✅|✅|✅|
|[ipfs](https://github.com/ipfs/js-ipfs/pull/830) |✅|✅|❌|❌|

@diasdavid @lgierth @whyrusleeping @dignifiedquire 

~~NOTE: I'm currently bringing this issue up to date with all the latest developments - this is still in WIP.~~

EDIT: This issue is up to date!

dryajov commented:
@diasdavid 

Here are the action items discussed last week:

- [x] Clean up libp2p tests 
- [x] Add swarm tests
- [ ] Work on go-js interop tests (WIP)
- [ ] Prepare demo that showcases circuit (WIP)
   - [ ] Use the files transfer example

dryajov commented:
A quick update on whats outstanding: 

- mafmt broke with the latest changes for webrtc addresses (mafmt master), I'm working on resolving it. The only address thats broken is the webrtc one, so it isn't really blocking the rest of the effort. 
- interop tests in js-ipfs. I've got some basic tests set up in test/core that work, but need to get additional interop tests set up:

|test|go-relay|js-relay|
|----|--------|--------| 
|go<->go| | |
|go<->js-cli| | |
|go<->js-browser| | |
|js-cli<->go| | |
|js-cli<->js-cli| | |
|js-cli<->js-browser| | |
|js-browser<->go| | |
|js-browser<->js-cli | |
|js-browser<->js-browser| | |

diasdavid commented:
As an update to everyone following this PR, check @dryajov demoing it here! https://youtu.be/chAXj_vsR2s?t=25m01s

dryajov commented:
This are the notes (from memory) of a discussion held between @vyzo @stebalien @diasdavid @dryajov on Sept. 14 regarding releasing go and js circuit:

Go:

- Go circuit is merged and awaiting go ipfs release - rc candidate is coming soon
- Go circuit does not announce circuit addresses to the DHT, but it would be nice to have a flag that would allow switching that functionality on/off - handy for experimenting once circuit is released.
   - @diasdavid should we add something like that to the JS implementation - probably not needed immediately since we're not announcing peers on the DHT just yet?

Js:
- Interop testing is happening - see table in https://github.com/ipfs/js-ipfs/pull/830#issuecomment-327253345.
- https://github.com/whyrusleeping/js-mafmt/pull/16 is no longer needed as it will be handled in the `filter` function of each transport.
- @diasdavid to review and release libp2p-circuit to npm
- @dryajov to finish up testing
- common goal is to have circuit ready for allhands next week (Sept. 18)

dryajov commented:
An update on the current state of circuit:

The good news:

- I've added most interop tests concerning go <-> js in the cli, browser is bit more complex, but I have a browser <-> circuit <-> browser test in core which passes
- The changes discussed during the last call and captured in the comment above have been taken care of and are reflected in the table in the first comment
- Manual testing was also done and the issues that were identified have been all taken care of
- In general coding is complete and things just need a good review

The not so good news:

Running tests with stuff linked locally breaks the test suit in several places, basically some tests will fail when ran as part of the whole suit but work when ran isolated, I'm not sure whats going on and I haven't yet tried recreating this inside a docker container, so it might be my setup or just the crazy linking thats going on underneath. Here are my raw notes of things that I've seen fail:

```
 - interop repo (test/interop/repo.js) tests are failing with latest go ipfs master...
 - Tests in interface-ipfs-core are failing i.e. - "Connecting two peers with one address each"
 - some before each hooks in core/files-sharding.spec.js ("should be able to add dir without sharding") fail when running as part of the suit
 - bootstrap.spec.js fails when running as part of the suit
- init.spec.js fails when ran as part of the suit
-  gateway tests are failing when ran as part of the suit
- bitswap.spec.js fails when ran as part of the suit
```

dryajov commented:
To help move this issue further, here is  break down of where things are for the reviewers:

- Coding is done across all modules and awaiting review
- Interop tests for the js-go on the cli, have been added under `test/interop`
- A simple browser test (browser<->relay<->browser) has also been added using an external node 
   - more complex browser tests are required (#1030)
- The table on the first comment reflects the current state of the PRs
- The issues mentioned in the previous comment are still present and being troubleshooted. _It would be nice to get a second set of eyes to eliminate npm linking issues_

diasdavid commented:
@dryajov started reviewing and merging, got blocked due to aegir update. Could you help with that too?

dryajov commented:
@diasdavid I'll update aegir to latest in circuit.

dryajov commented:
From a conversation with @diasdavid, here is what needs to happen next to get circuit into js-ipfs!

The bellow modules need to be released first as everything else depends on them:

- [x] **libp2p-tcp** and **libp2p-webrtc-direct** have been released. yay! ❤️ 
- [ ] **libp2p-websockets**
  - updated to latest aegir and awaiting review/release
- [ ] **libp2p-webrtc-start** 
  - updated to latest aegir and awaiting review/release

Once the above modules are released, we can start moving the next batch in the order listed:

- [ ] **libp2p-circuit** 
  - updated to the newest aegir and awaiting a review from @VictorBjelkholm ⭐️ 
- [ ] **libp2p-swarm** 
  - upgraded to the newest aegir and is awaiting **libp2p-circuit** release
- [ ] **libp2p** 
  - upgraded to the newest aegir and awaiting **libp2p-swarm**
- [ ] **js-ipfs** 
   - [ ] needs updating to newest aegir
   - [ ] needs **libp2p** released

diasdavid commented:
@dryajov what's the update on this https://github.com/ipfs/js-ipfs/pull/830#issuecomment-327253345 ?

dryajov commented:
@diasdavid I'll update that table shortly.

diasdavid commented:
@dryajov libp2p released, this PR is fully unblocked :)

dryajov commented:
@diasdavid I got the tests passing locally, but we need - https://github.com/ipfs/aegir/pull/166 and https://github.com/libp2p/js-libp2p-swarm/pull/234. I think we're getting closer 😉 🍹

diasdavid commented on package.json:
```diff
@@ -74,7 +74,7 @@
     "form-data": "^2.3.1",
     "gulp": "^3.9.1",
-    "interface-ipfs-core": "~0.32.1",
+    "interface-ipfs-core": "^0.33.1",
```
@dryajov don't flip the semver from ~ to ^

dryajov commented:
@diasdavid do we have any project wide guidelines as to when we want to use ^ or ~? Semver leaves it up to the project/org to decide what works best. I thought that at some point we moved most things to ^? 

Just curious, thanks in advance!

diasdavid commented:
@dryajov We always use ~ when it is < 1.0.0. I'll try to find the notes we wrote on why. Essentially we use Minor to signal breaking changes while in pre 1.0.0.

dryajov commented:
ah, that makes sense! thanks!

diasdavid commented:
@dryajov what's the status on https://github.com/ipfs/js-ipfs/pull/830#issuecomment-327253345 ? Do you know why some of those tests are failing? 

Are there interop tests that test all of that?

dryajov commented:
@diasdavid re: https://github.com/ipfs/js-ipfs/pull/830#issuecomment-327253345, those tests are not implemented because coordinating starting and stopping nodes is currently very cumbersome. I'll add those tests in if absolutely necessary for this release, but if not, I would like to get something in place that allows starting nodes from browsers in a morey easy way.

dryajov commented:
@diasdavid tests are failing because we need to release the latest aegir master.

diasdavid commented on package.json:
```diff
@@ -19,17 +19,21 @@
+    "test:unit:node:core": "TEST=core aegir test --target node --no-parallel",
+    "test:unit:node:http": "TEST=http aegir test --target node --no-parallel  ",
+    "test:unit:node:gateway": "TEST=gateway aegir test --target node --no-parallel",
+    "test:unit:node:cli": "TEST=cli aegir test --target node --no-parallel",
```
With the new AEgir, you should be able to do something better than setting an env var. See https://github.com/ipfs/aegir/issues/135

diasdavid replied:
@dryajov are you on it?

diasdavid commented on package.json:
```diff
@@ -103,43 +103,43 @@
+    "ipfs-repo": "^0.17.0",
+    "ipfs-unixfs": "^0.1.13",
+    "ipfs-unixfs-engine": "^0.22.5",
+    "ipld-resolver": "^0.13.4",
```
@dryajov please do not swap ~ for ^

diasdavid commented on src/core/components/libp2p.js:
```diff
@@ -20,7 +20,14 @@ module.exports = function libp2p (self) {
+            enabled: get(config, 'EXPERIMENTAL.Relay.Enabled', false),
+            hop: {
+              enabled: get(config, 'EXPERIMENTAL.Relay.HOP.Enabled', false),
+              active: get(config, 'EXPERIMENTAL.Relay.HOP.Active', false)
```
What is the difference between Enabled and Active?

dryajov replied:
`Active` is for active/passive relaying, `Enabled` is just the flag.

diasdavid replied:
Is this matching go-ipfs config?

diasdavid commented on test/core/circuit.spec.js:
```diff
@@ -0,0 +1,172 @@
+      }
+    }
+  })
+}
```
Why not use the default utils?

diasdavid replied:
Please avoid creating so much duplicated code for creating nodes (the same thing happened in libp2p). Always create code that is reusable and of low maintenance.

dryajov commented on .aegir.js:
```diff
@@ -1,12 +1,146 @@
+    (cb) => spawnGoNode(31, cb),
+    (cb) => spawnGoNode(32, cb)
+  ], done)
+}
```
sure, I'll move them over.

dryajov commented on .aegir.js:
```diff
@@ -1,12 +1,146 @@
+  ], done)
+}
+
+let after = (done) => {
```
good catch, should be `const`.

diasdavid commented:
![image](https://user-images.githubusercontent.com/1211152/32365422-b5cada7a-c071-11e7-9178-f07db44cbd4f.png)

What's missing in this PR?

diasdavid commented on .aegir.js:
```diff
@@ -1,12 +1,60 @@
+      `/ip4/127.0.0.1/tcp/20032/ws`
+    ], 33032, 44032, cb)
+  ], done)
+}
```
Why can't the "Spawner" use IPFS Factory?

diasdavid replied:
Also, what are the magic numbers?

dryajov replied:
The numbers are the preexisting port numbers that this daemons were being started with - I kept them because I believe some other tests might rely on them.

diasdavid commented on package.json:
```diff
@@ -144,26 +144,26 @@
   "optionalDependencies": {
     "prom-client": "^10.2.0",
-    "prometheus-gc-stats": "^0.5.0"
+    "prometheus-gc-stats": "~0.5.0"
```
Thank you :)

diasdavid commented on test/cli/bootstrap.js:
```diff
@@ -8,7 +8,8 @@ const runOnAndOff = require('../utils/on-and-off')
 
-  before(() => {
+  before(function () {
+    // this.timeout(30 * 1000)
```
If there is no need for the timeout, revert back to an arrow function and remove the comment

dryajov replied:
@diasdavid yeah, I'm getting some random test fails and trying to isolate as to whats causing them, this commented timeouts are a WIP, disregard for now please.

diasdavid commented on test/cli/commands.js:
```diff
@@ -9,7 +9,8 @@ const commandCount = 56
 
-  before(() => {
+  before(function () {
+    // this.timeout(30 * 1000)
```
Same here

diasdavid commented on test/cli/id.js:
```diff
@@ -7,7 +7,8 @@ const runOnAndOff = require('../utils/on-and-off')
 
-  before(() => {
+  before(function () {
+    // this.timeout(30 * 1000)
```
Same here

diasdavid commented on test/cli/init.js:
```diff
@@ -40,7 +40,8 @@ describe('init', () => {
     }).then((out) => expect(out).to.equal(readme))
-  }).timeout(8000)
+  })
+    // .timeout(20 * 1000)
```
If the timeout is not needed, just delete

diasdavid commented on test/cli/version.js:
```diff
@@ -8,7 +8,8 @@ const runOnAndOff = require('../utils/on-and-off')
 
-  before(() => {
+  before(function () {
+    // this.timeout(30 * 1000)
```
same here

diasdavid commented on test/core/bitswap.spec.js:
```diff
@@ -39,7 +38,8 @@ function makeBlock (cb) {
 
-  beforeEach((done) => {
+  beforeEach(function (done) {
+    // this.timeout(20 * 1000)
```
same here

diasdavid commented on test/core/bitswap.spec.js:
```diff
@@ -144,7 +144,7 @@ describe('bitswap', () => {
 
       it('3 peers', function (done) {
-        this.timeout(60 * 1000)
+        // this.timeout(60 * 1000)
```
If the timeout is not needed, convert to arrow function and remove the comment

diasdavid commented on test/core/bitswap.spec.js:
```diff
@@ -221,7 +221,8 @@ describe('bitswap', () => {
 
-    before((done) => {
+    before(function (done) {
+      // this.timeout(20 * 1000)
```
Same here

diasdavid commented on test/core/bitswap.spec.js:
```diff
@@ -262,6 +263,7 @@ describe('bitswap', () => {
       it('.wantlist returns an array of wanted blocks', () => {
         expect(node.bitswap.wantlist()).to.eql([])
       })
+        // .timeout(30 * 1000)
```
Same here

diasdavid commented on test/core/circuit.spec.js:
```diff
@@ -0,0 +1,140 @@
+          (cb) => peerInfoFromObj(res2[0], cb),
+          (cb) => peerInfoFromObj(res2[1], cb),
+          (cb) => peerInfoFromObj(res2[2], cb),
+          (cb) => peerInfoFromObj(res2[3], cb)
```
Why do you need a peerInfo object? node.id() returns an array of multiaddrs with the IPFS peerId, those should be enough for a dial.

dryajov replied:
That's because the peer object returned from the API doesn't match the `PeerInfo` structure exactly. This is what I get when calling `peer.id`

```
  {
    "id": "QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC",
    "publicKey": "CAASpgIwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCS+u+s1cQ0HYHkVLzf8MIS4TaZq9TEl0mezvadcn2qGiOGcOMBAZ+jtH3Mcy9Bf2Y/Wozd/8lJaHG7Exnwlfxmb82vSEW2IAc7UoULG7MkUQdSC5pq7plq/RkMhzMHmiXjsnqkPdpSIi7DfNYgiNYIpd4f455pvMgcjvlpZI5BHfZtssydSkQmaUxUQughdJCqFa8RhVb4pZUEGkK7y6a286pi8NH3DXPsc3FLd6TTJOHa0T7DxErpe9+XoVN5k+mY7KbPMnnc3KhKKfEHAiRwCssqI3BTlk60k0Qet7FnYexdcdQ3d8tBkJwU9DpW/ugCUyLpytrsmEI5Yml40E2ZAgMBAAE=",
    "addresses": [
      "/ip4/127.0.0.1/tcp/10008/ipfs/QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC",
      "/ip4/127.0.0.1/tcp/20008/ws/ipfs/QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC",
      "/p2p-circuit/ip4/127.0.0.1/tcp/10008/ipfs/QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC",
      "/p2p-circuit/ip4/127.0.0.1/tcp/20008/ws/ipfs/QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC",
      "/p2p-circuit/ipfs/QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC"
    ],
    "agentVersion": "js-ipfs",
    "protocolVersion": "9000"
  }
```

I believe that the `addresses` prop is now expected to be called `multiaddrs`, that makes it fail with `Error: peer type not recognized]` in libp2p in the `_getPeerInfo` helper. I think this is something that needs to be fixed in the api itself, but I wasn't aware of it being a problem until looking into it now.

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -37,6 +37,7 @@ describe('create node', () => {
       })
     })
   })
+    // .timeout(10 * 1000)
```
If not needed then remove

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -59,6 +60,7 @@ describe('create node', () => {
       })
     })
   })
+    // .timeout(10 * 1000)
```
Same here

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -84,6 +86,7 @@ describe('create node', () => {
       })
     })
   })
+    // .timeout(20 * 1000)
```
One more

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -173,6 +176,7 @@ describe('create node', () => {
 
     node.once('ready', () => node.start())
   })
+    // .timeout(10 * 1000)
```
Another one

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -192,6 +196,7 @@ describe('create node', () => {
       node.start(() => node.stop(done))
     })
   })
+    // .timeout(10 * 1000)
```
Uno mas

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -222,6 +227,7 @@ describe('create node', () => {
       })
     })
   })
+    // .timeout(10 * 1000)
```
plus one

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -241,6 +247,7 @@ describe('create node', () => {
       (cb) => node.stop(cb)
     ], done)
   })
+    // .timeout(8000)
```
target: eliminate all timeouts

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -264,4 +271,5 @@ describe('create node', () => {
       (cb) => node.stop(cb)
     ], done)
   })
+    // .timeout(10 * 1000)
```
destroy all the commented timeouts

diasdavid commented on test/core/files-sharding.spec.js:
```diff
@@ -61,7 +60,8 @@ describe('files dir', () => {
 
-    before((done) => {
+    before(function (done) {
+      // this.timeout(10 * 1000)
```
🔫 pew pew pew

diasdavid commented on test/core/init.spec.js:
```diff
@@ -52,7 +52,8 @@ describe('init', () => {
 
-  it('set # of bits in key', (done) => {
+  it('set # of bits in key', function (done) {
+    // this.timeout(10 * 1000)
```
💣

diasdavid commented on test/gateway/index.js:
```diff
@@ -25,7 +25,8 @@ describe('HTTP Gateway', () => {
 
-  before((done) => {
+  before(function (done) {
+    // this.timeout(20 * 1000)
```
🔥

diasdavid commented on test/gateway/index.js:
```diff
@@ -250,3 +251,4 @@ describe('HTTP Gateway', () => {
     })
   })
 })
+  // .timeout(10 * 1000)
```
🤺

diasdavid commented on test/interop/circuit/index.js:
```diff
@@ -0,0 +1,209 @@
+  let jsWS
+  let jsWSAddrs
+  let jsRelayAddrs
+  let factory = new Factory()
```
Can't this be const?

diasdavid commented on test/interop/circuit/index.js:
```diff
@@ -0,0 +1,209 @@
+
+chai.use(dirtyChai)
+
+describe('circuit interop', function () {
```
Can this be arrow func?

diasdavid commented on test/interop/circuit/index.js:
```diff
@@ -0,0 +1,209 @@
+
+  beforeEach((done) => {
+    parallel([
+      (pCb) => utils.setupJsNode([
```
Please require this util above directly and use it as a single func.

Is there a way that setupJsNode can be replaced by ipfs-factory?

diasdavid commented on test/utils/on-and-off.js:
```diff
@@ -16,14 +16,16 @@ function off (tests) {
 
-    before(() => {
+    before(function () {
+      // this.timeout(30 * 1000)
```
No need for timeouts! (JEST is gone \o/)

diasdavid commented on test/utils/on-and-off.js:
```diff
@@ -16,14 +16,16 @@ function off (tests) {
 
-    after((done) => {
+    after(function (done) {
+      // this.timeout(20 * 1000)
```
🏏

diasdavid commented on test/utils/on-and-off.js:
```diff
@@ -41,7 +43,7 @@ function on (tests) {
       // so we need to increase the timeout for the
       // before step
-      this.timeout(20 * 1000)
+      // this.timeout(30 * 1000)
```
If not needed, then it is not needed

diasdavid commented on test/utils/on-and-off.js:
```diff
@@ -53,7 +55,10 @@ function on (tests) {
+    after(function (done) {
+      // this.timeout(20 * 1000)
+      factory.dismantle(done)
+    })
```
Bring back the one liner

diasdavid commented on test/utils/spawn-tools.js:
```diff
@@ -0,0 +1,110 @@
+  spawnNodeDaemon,
+  spawnGoNode,
+  stopNodes
+}
```
Can't all of this be replaced by ipfs-factory? I bet it can :)

diasdavid commented:
Seems that Travis and Circle have different tests failing:

Circle:
![image](https://user-images.githubusercontent.com/1211152/32366046-5a02cc5e-c074-11e7-8ad0-a3422620915e.png)

Travis:
![image](https://user-images.githubusercontent.com/1211152/32366036-4de9a26c-c074-11e7-9b0e-e981ae1900c6.png)

My spidey sense tells me that they are falling in waterfall (one fails, doesn't close the daemon well and the others have a clash)

dryajov commented:
@diasdavid yeah, the random tests failures are what I'm trying to figure out right now. The thing is that a different set fails locally, which makes it a lot of fun to track down.

dryajov commented:
@diasdavid the ETA is as soon as I can get the tests passing and cleanup the issues you highlighted - I'm shooting for the weekend or early next week.

dryajov commented:
Here is a summary of the outstanding things:

- [ ] figure out the random test failures
- [ ] do some clean up in tests/node creation
- [ ] normalize config across go and js (use go's configs as they have already shipped)
- [ ] release by Nov ~6th.

diasdavid commented:
@dryajov Let's agree that if by tomorrow this PR is still not completely finished. We .skip the Relay tests and ensure all the other ones are passing, so that master gets refreshed with Aegir and all the code that has been shipped and then another PR is created to fix the rest of the Relay issues.

dryajov commented:
@diasdavid that sounds reasonable. I'll tidy this up so that its mergeable.

dryajov commented:
BTW, the interop tests when ran with `npm run test:interop` which run against both js and go nodes, actually pass. I suspect that these tests failing has something to do with attaching to an external go-ipfs node over the http API from the mocha tests - attaching to an external js-ipfs node works fine.

diasdavid commented on package.json:
```diff
@@ -21,13 +21,13 @@
-    "test:browser": "aegir test --target browser --target webworker --no-parallel --no-cors --timeout 100000",
+    "test": "aegir test --target node --target browser --no-cors --no-parallel",
+    "test:node": "aegir test --target node --no-parallel",
+    "test:browser": "aegir test --target browser --target webworker --no-parallel --no-cors",
```
You can actually take the `--no-parallel` out since we removed JEST

diasdavid commented on test/core/circuit.spec.js:
```diff
@@ -138,5 +138,5 @@ describe('circuit', function () {
       ], done)
     })
-  })
+  }).timeout(20 * 1000)
```
Please add the timeouts at the top of the file just for consistency.

diasdavid commented on package.json:
```diff
@@ -21,19 +21,19 @@
+    "test:unit:node:core": "aegir test --target node -f test/core/*.js",
+    "test:unit:node:http": "aegir test --target node -f test/http-api/index.js",
+    "test:unit:node:gateway": "aegir test --target node -f test/gateway/index.js",
+    "test:unit:node:cli": "aegir test --target node -f test/cli/index.js",
```
Woot! Love this :D

dryajov replied:
😜 getting there. baby steps :)

dryajov commented:
@diasdavid I've got both circle and travis tests passing. The only test that's currently disabled is the go-ipfs circuit tests that is mainly there for browser runs. It hangs for some unknown reason in mocha runs, but works fine when testing manually. 

I think this is ready to proceed with the original plan of merging and handling the remaining problems in separate issues.

diasdavid commented on .aegir.js:
```diff
@@ -1,12 +1,71 @@
+
+const spawnJsNode = require('./test/utils/spawn-tools').spawnJsNode
+const spawnGoNode = require('./test/utils/spawn-tools').spawnGoNode
+const stopNodes = require('./test/utils/spawn-tools').stopNodes
```
Require once, use multiple times

diasdavid commented on test/core/bitswap.spec.js:
```diff
@@ -187,12 +188,12 @@ describe('bitswap', () => {
-      })
-    })
+      }).timeout(20 * 1000)
+    }).timeout(60 * 1000)
```
Move timeouts up

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -36,7 +36,7 @@ describe('create node', () => {
       })
     })
-  })
+  }).timeout(20 * 1000)
```
move timeouts up

dryajov replied:
I've seen both ways being used. If we want to stick to timeouts at the top, I think we should switch to using `function` instead of `=>`.

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -58,7 +58,7 @@ describe('create node', () => {
       })
     })
-  })
+  }).timeout(20 * 1000)
```
move timeouts up

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -83,7 +83,7 @@ describe('create node', () => {
       })
     })
-  })
+  }).timeout(20 * 1000)
```
move timeouts up

diasdavid commented on test/core/create-node.spec.js:
```diff
@@ -172,7 +172,7 @@ describe('create node', () => {
 
     node.once('ready', () => node.start())
-  })
+  }).timeout(20 * 1000)
```
move timeouts up

diasdavid commented on .aegir.js:
```diff
@@ -1,12 +1,71 @@
+    (cb) => spawnGoNode([
+      `/ip4/127.0.0.1/tcp/10032`,
+      `/ip4/127.0.0.1/tcp/20032/ws`
+    ], true, 33032, 44032, cb)
```
This is spawning a ton of go nodes for the interop tests, but interop tests never run this script because they use mocha directly

dryajov replied:
True, but interop tests now rely nodes spawned within the tests them self, do we want to rework it in this PR?

diasdavid commented on package.json:
```diff
@@ -19,17 +19,21 @@
+    "test:unit:browser": "aegir test --target browser --no-cors",
     "test:interop": "npm run test:interop:node",
     "test:interop:node": "mocha -t 60000 test/interop/node.js",
     "test:interop:browser": "mocha -t 60000 test/interop/browser.js",
```
This needs to start using aegir due to https://github.com/ipfs/js-ipfs/pull/830/files#r149579164

dryajov replied:
I'll get those in.

diasdavid commented on .aegir.js:
```diff
@@ -1,12 +1,43 @@
+    pre: pre,
+    post: stop
   }
 }
```
@dignifiedquire is there a way to select custom .aegir.js for different runs? Goal here is to have one for interop and another for unit

dryajov replied:
AFAIK, we can have `browser.hooks`, but not sure if we can have custom `tags`.

diasdavid commented:
@dryajov please go ahead and create a PR enabling Circuit Relay interop tests

diasdavid commented:
The work continues on https://github.com/ipfs/js-ipfs/pull/1063
